applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp

```diff
@@ -1015,6 +1021,36 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAll(MatrixType&        rLe
     KRATOS_CATCH("")
 }
 
+template <unsigned int TDim, unsigned int TNumNodes>
+std::vector<double> UPwSmallStrainElement<TDim, TNumNodes>::CalculateDerivativesOfSaturation(
+    const std::vector<double>& fluid_pressures, RetentionLaw::Parameters& RetentionParameters)
+{
+    std::vector<double> result;
+    std::transform(fluid_pressures.begin(), fluid_pressures.end(), mRetentionLawVector.begin(),
+                   std::back_inserter(result),
+                   [&RetentionParameters](double fluid_pressure, RetentionLaw::Pointer pRetentionLaw) {
+        RetentionParameters.SetFluidPressure(fluid_pressure);
+        return pRetentionLaw->CalculateDerivativeOfSaturation(RetentionParameters);
+    });
+
+    return result;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+std::vector<double> UPwSmallStrainElement<TDim, TNumNodes>::CalculateDegreesOfSaturation(
+    const std::vector<double>& fluid_pressures, RetentionLaw::Parameters& RetentionParameters)
+{
+    std::vector<double> result;
+    std::transform(fluid_pressures.begin(), fluid_pressures.end(), mRetentionLawVector.begin(),
+                   std::back_inserter(result),
+                   [&RetentionParameters](double fluid_pressure, RetentionLaw::Pointer pRetentionLaw) {
+        RetentionParameters.SetFluidPressure(fluid_pressure);
+        return pRetentionLaw->CalculateSaturation(RetentionParameters);
+    });
+
+    return result;
+}
```

Comment:
It's a shame this is a 100% duplication with diff order. We could extract a utility, but we'd need to supply the vector of retention laws. It's still open for discussion though, I'm doubting between the two possibilities

Reply:
As you mentioned, this part is also duplication, and it has to be shifted to a utility.